src/librustc_lint/unused.rs

```diff
@@ -355,7 +388,50 @@ impl UnusedParens {
                         left_pos.map(|s| s >= value.span.lo()).unwrap_or(false),
                         right_pos.map(|s| s <= value.span.hi()).unwrap_or(false),
                     );
-                    Self::remove_outer_parens(cx, value.span, &expr_text, msg, keep_space);
+                    Self::emit_unused_parens(cx, value.span, &expr_text, ctx.into(), keep_space);
+                }
+            }
+            ast::ExprKind::Block(ref inner, None)
+                if inner.rules == ast::BlockCheckMode::Default =>
+            {
+                // emit a warning under the following conditions:
+                //
+                // - the block does not have a label
+                // - the block is not `unsafe`
+                // - the block contains exactly one expression (do not lint `{ expr; }`)
+                // - `followed_by_block` is true and the internal expr may contain a `{`
+                // - the block is not multiline
+                // - the block has no attribute and was not created inside a macro
+                // - if the block is a constant expression, the inner expr must be a literal.
```

Comment:
and the rationale for this?

Reply:
because of const generics, which sometimes require blocks, i.e. we don't want to lint

```rust
#![feature(const_generics)]
//~^ WARN the feature `const_generics` is incomplete and may cause the compiler to crash

trait IsZeroTrait<const IS_ZERO: bool>{}

impl IsZeroTrait<{0u8 == 0u8}> for () {} // the block is required,
// impl IsZeroTrait<0u8 == 0u8> // this does not compile

impl IsZeroTrait<true> for ((),) {}

fn main() {}
```

Reply:
Please extend the comment to mention this, but also, having a path works fine in a block wrt. const generics.

Reply:
updated the comment to `is a anon const`.

I don't know how to check whether the anon const is an array length or a generic param.
But I also don't think that being too conservative is a big problem.

> but also, having a path works fine in a block wrt. const generics.

Sadly not, see https://github.com/rust-lang/rust/pull/70081#issuecomment-600496770

Reply:
added comments explaining the rationale.